Flakiness in core2.test_pthread_proxying_refcount

See https://app.circleci.com/pipelines/github/emscripten-core/emscripten/29174/workflows/e3fe5072-068b-4775-9d62-0a3bbf0c8ad2/jobs/657474

```
AssertionError: JS subprocess failed (/root/emsdk/node/16.20.0_64bit/bin/node --stack-trace-limit=50 --trace-uncaught /tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js): 7.  Output:
start
Aborted(Assertion failed: frees_after_cull > frees_before_cull, at: /root/project/test/pthread/test_pthread_proxying_refcount.c,100,main)
/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1
var Module=typeof Module!="undefined"?Module:{};var moduleOverrides=Object.assign({},Module);var arguments_=[];var thisProgram="./this.program";var quit_=(status,toThrow)=>{throw toThrow};var ENVIRONMENT_IS_WEB=typeof window=="object";var ENVIRONMENT_IS_WORKER=typeof importScripts=="function";var ENVIRONMENT_IS_NODE=typeof process=="object"&&typeof process.versions=="object"&&typeof process.versions.node=="string";var ENVIRONMENT_IS_PTHREAD=Module["ENVIRONMENT_IS_PTHREAD"]||false;var _scriptDir=[..]
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    [..]

RuntimeError: Aborted(Assertion failed: frees_after_cull > frees_before_cull, at: /root/project/test/pthread/test_pthread_proxying_refcount.c,100,main). Build with -sASSERTIONS for more info.
    at abort (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:7649)
    at ___assert_fail (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:17949)
    at wasm://wasm/00015f42:wasm-function[23]:0xb6c
    at Module._main (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:24517)
    at callMain (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:27699)
    at doRun (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:28138)
    at run (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:28306)
    at runCaller (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:27568)
    at removeRunDependency (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:7455)
    at /tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:13724
```

--- a/test/pthread/test_pthread_proxying_refcount.c
+++ b/test/pthread/test_pthread_proxying_refcount.c
@@ -37,7 +37,9 @@ void register_processed(void) {
   processed++;
 }
 
-void task(void* arg) { *(_Atomic int*)arg = 1; }
+void set_flag(void* arg) { *(_Atomic int*)arg = 1; }
+
+void task(void* arg) { emscripten_async_call(set_flag, arg, 0); }
 
 void* execute_and_free_queue(void* arg) {
   // Wait until we are signaled to execute the queue.
@@ -74,16 +76,10 @@ int main() {
   }
   should_execute = 1;
 
-  // Wait for the tasks to be executed.
+  // Wait for the tasks to be executed. The queues will have been destroyed
+  // after this.
   while (!executed[0] || !executed[1]) {
   }
-
-  // Wait a bit (20 ms) for the notification to be received.
-  struct timespec time = {
-    .tv_sec = 0,
-    .tv_nsec = 20 * 1000 * 1000,
-  };
-  nanosleep(&time, NULL);
 
 #ifndef SANITIZER
   // Our zombies should not have been freed yet.

--- a/test/pthread/test_pthread_proxying_refcount.c
+++ b/test/pthread/test_pthread_proxying_refcount.c
@@ -1,9 +1,10 @@
 #include <assert.h>
 #include <emscripten/console.h>
 #include <emscripten/emscripten.h>
-#include <emscripten/proxying.h>
 #include <emscripten/heap.h>
+#include <emscripten/proxying.h>
 #include <pthread.h>
+#include <sched.h>
 #include <stdbool.h>
 #include <unistd.h>
 
@@ -39,11 +40,14 @@ void register_processed(void) {
 
 void set_flag(void* arg) { *(_Atomic int*)arg = 1; }
 
+// Delay setting the flag until the next turn of the event loop so it can be set
+// after the proxying queue is destroyed.
 void task(void* arg) { emscripten_async_call(set_flag, arg, 0); }
 
 void* execute_and_free_queue(void* arg) {
   // Wait until we are signaled to execute the queue.
   while (!should_execute) {
+    sched_yield();
   }
 
   // Execute the proxied work then free the empty queues.
@@ -79,6 +83,7 @@ int main() {
   // Wait for the tasks to be executed. The queues will have been destroyed
   // after this.
   while (!executed[0] || !executed[1]) {
+    sched_yield();
   }
 
 #ifndef SANITIZER

